### **This section is used to scrape the provided wikipedia page to produce the table of zip codes here in this notebook using pandas:**

*Note: I had to run this library install as separate cells since I encountered so many errors

In [62]:
import numpy as np # library to handle data in a vectorized manner

In [63]:
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [64]:
import json # library to handle JSON files

In [67]:
!pip install geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
print('imported.')


Usage:   
  pip install [options] <requirement specifier> [package-index-options] ...
  pip install [options] -r <requirements file> [package-index-options] ...
  pip install [options] [-e] <vcs project url> ...
  pip install [options] [-e] <local project path> ...
  pip install [options] <archive url/path> ...

no such option: --yes
imported.


In [68]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [69]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [70]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [73]:
!pip install folium # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
print('imported.')

     |████████████████████████████████| 102kB 19.4MB/s ta 0:00:01
imported.


In [74]:
df=pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")[0]

Here is the initial dataframe before I make any changes:

In [75]:
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

Next, I will remove rows where no borogh has been assigned:

In [76]:
df_clean = df[df.Borough != 'Not assigned']
df_clean

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Vic

The assignment then states: "If a cell has a borough but a **Not assigned** neighborhood, then the neighborhood will be the same as the borough." but the table above does not appear to have any boroughs where the neighborhood is **Not assigned** so I do not see any need to edit the neighborhood column to read the same as the borough name for that row. Looking through the initial dataframe as well as the initial table on the wikipedia page, it appears there are no neighborhoods with missing information unless the borogh name is also **Not assigned**, and I've already removed those.

In [77]:
df_clean.shape


(103, 3)

It is worth noting that the instructions for this assignement say the dataframe will consist of three columns. Base on running .shape above we can see that my df_clean dataframe indeed has 3 columns, even though the display technically has 4 including the leading index. The presentation in the assignment also had a leading index column so I did not remove this.

### This section begins the 2nd part of the week 3 assignment where we gather the latitude & longitude coordinates for the postal codes in our clean dataframe from step one above


****Note that I tried the geocoder.google process as outlined in the instructions which ran into error after error. I then tried the "!conda install -c conda-forge geocoder --yes" process and still encountered error after error. Finally I decided to just use the csv link provided

In [78]:
df_latlng = pd.read_csv('https://cocl.us/Geospatial_data')

In [39]:
df_latlng

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

I will now merge the df_clean dataframe from step one with this csv file I named df_latlng

In [79]:
df_merged = pd.merge(df_latlng, df_clean, on='Postal Code')

In [45]:
df_merged

Postal Code   Latitude  Longitude           Borough  \
0           M1B  43.806686 -79.194353       Scarborough   
1           M1C  43.784535 -79.160497       Scarborough   
2           M1E  43.763573 -79.188711       Scarborough   
3           M1G  43.770992 -79.216917       Scarborough   
4           M1H  43.773136 -79.239476       Scarborough   
5           M1J  43.744734 -79.239476       Scarborough   
6           M1K  43.727929 -79.262029       Scarborough   
7           M1L  43.711112 -79.284577       Scarborough   
8           M1M  43.716316 -79.239476       Scarborough   
9           M1N  43.692657 -79.264848       Scarborough   
10          M1P  43.757410 -79.273304       Scarborough   
11          M1R  43.750072 -79.295849       Scarborough   
12          M1S  43.794200 -79.262029       Scarborough   
13          M1T  43.781638 -79.304302       Scarborough   
14          M1V  43.815252 -79.284577       Scarborough   
15          M1W  43.799525 -79.318389       Scarborough   
16          M1X  43.836125 -79.205636       Scarborough   
17          M2H  43.803762 -79.363452        North York   
18          M2J  43.778517 -79.346556        North York   
19          M2K  43.786947 -79.385975        North York   
20          M2L  43.757490 -79.374714        North York   
21          M2M  43.789053 -79.408493        North York   
22          M2N  43.770120 -79.408493        North York   
23          M2P  43.752758 -79.400049        North York   
24          M2R  43.782736 -79.442259        North York   
25          M3A  43.753259 -79.329656        North York   
26          M3B  43.745906 -79.352188        North York   
27          M3C  43.725900 -79.340923        North York   
28          M3H  43.754328 -79.442259        North York   
29          M3J  43.767980 -79.487262        North York   
..          ...        ...        ...               ...   
73          M6C  43.693781 -79.428191              York   
74          M6E  43.689026 -79.453512              York   
75          M6G  43.669542 -79.422564  Downtown Toronto   
76          M6H  43.669005 -79.442259      West Toronto   
77          M6J  43.647927 -79.419750      West Toronto   
78          M6K  43.636847 -79.428191      West Toronto   
79          M6L  43.713756 -79.490074        North York   
80          M6M  43.691116 -79.476013              York   
81          M6N  43.673185 -79.487262              York   
82          M6P  43.661608 -79.464763      West Toronto   
83          M6R  43.648960 -79.456325      West Toronto   
84          M6S  43.651571 -79.484450      West Toronto   
85          M7A  43.662301 -79.389494  Downtown Toronto   
86          M7R  43.636966 -79.615819       Mississauga   
87          M7Y  43.662744 -79.321558      East Toronto   
88          M8V  43.605647 -79.501321         Etobicoke   
89          M8W  43.602414 -79.543484         Etobicoke   
90          M8X  43.653654 -79.506944         Etobicoke   
91          M8Y  43.636258 -79.498509         Etobicoke   
92          M8Z  43.628841 -79.520999         Etobicoke   
93          M9A  43.667856 -79.532242         Etobicoke   
94          M9B  43.650943 -79.554724         Etobicoke   
95          M9C  43.643515 -79.577201         Etobicoke   
96          M9L  43.756303 -79.565963        North York   
97          M9M  43.724766 -79.532242        North York   
98          M9N  43.706876 -79.518188              York   
99          M9P  43.696319 -79.532242         Etobicoke   
100         M9R  43.688905 -79.554724         Etobicoke   
101         M9V  43.739416 -79.588437         Etobicoke   
102         M9W  43.706748 -79.594054         Etobicoke   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5  

Finally I will rearrange the columns to match the table shown in the assignment, put the result into a new dataframe called df_reorder, and display the header using 10 rows:

In [80]:
df_reorder = df_merged.reindex(columns=['Postal Code', 'Borough', 'Neighborhood', 'Latitude', 'Longitude'])
df_reorder.head(10)

Postal Code      Borough                                     Neighborhood  \
0         M1B  Scarborough                                   Malvern, Rouge   
1         M1C  Scarborough           Rouge Hill, Port Union, Highland Creek   
2         M1E  Scarborough                Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                           Woburn   
4         M1H  Scarborough                                        Cedarbrae   
5         M1J  Scarborough                              Scarborough Village   
6         M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park   
7         M1L  Scarborough                  Golden Mile, Clairlea, Oakridge   
8         M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West   
9         M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

And the I will check the shape one more time to make sure we are still at (103 X 5)

In [81]:
df_reorder.shape

(103, 5)

### This section is the final (3rd) section of the assignment where we walk through performing the same exercise exploring & clustering neighborhoods in Toronto, just like we did in the previous New York lab exercise

Picking up where the New York exercise has a dataframe of boroughs & neighborhoods with latitude & longitude:

#### Use geopy library to get the latitude and longitude values of New York City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>to_explorer</em>, as shown below.

In [82]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Greater Toronto with ALL neighborhoods superimposed on top.

In [86]:
#  I copied some of the text from the New York exercise which relied on a dataframe named neighborhoods so I am creating a dataframe here called neighborhoods off of the df_reorder created previously
neighborhoods = df_reorder
neighborhoods.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [94]:
# create map of Greater Toronto using latitude and longitude values
map_great_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_great_toronto)  
    
map_great_toronto

For illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Toronto. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [91]:
toronto_data = neighborhoods[neighborhoods['Borough'].astype(str).str.contains('Toronto')]
toronto_data

Postal Code           Borough  \
37         M4E      East Toronto   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
87         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                  North Toronto West, Lawrence Park  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
66                     University of Toronto, Harbord  43.662696 -79.400049  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
69                                     Stn A PO Boxes  43.646435 -79.374846  
70             First Canadian Place, Underground city  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564  
76                       Dufferin, Dovercourt Village  43.669005 -

Now that we have a dataframe of just Toronto specific boroughs, lets visualize just those boroghs on a map:

In [105]:
# create map of Manhattan using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [120]:
CLIENT_ID = 'MLTOHZTUFRKGHKUQR4O2PMTX3ANCOZ3XMLSUC2VNFJMISJ4X' # your Foursquare ID
CLIENT_SECRET = 'HOHDJNMA5NRIFF523PDNQX2LDE2G24VHYC4HBRMFAO0H51SQ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MLTOHZTUFRKGHKUQR4O2PMTX3ANCOZ3XMLSUC2VNFJMISJ4X
CLIENT_SECRET:HOHDJNMA5NRIFF523PDNQX2LDE2G24VHYC4HBRMFAO0H51SQ


#### Let's explore the first neighborhood in our dataframe.

First, I realize the index is goofy after stripping down to just Toronto boroughs, so we need to reindex:

In [122]:
toronto_data = toronto_data.reset_index()

In [123]:
toronto_data

index Postal Code           Borough  \
0       0         M4E      East Toronto   
1       1         M4K      East Toronto   
2       2         M4L      East Toronto   
3       3         M4M      East Toronto   
4       4         M4N   Central Toronto   
5       5         M4P   Central Toronto   
6       6         M4R   Central Toronto   
7       7         M4S   Central Toronto   
8       8         M4T   Central Toronto   
9       9         M4V   Central Toronto   
10     10         M4W  Downtown Toronto   
11     11         M4X  Downtown Toronto   
12     12         M4Y  Downtown Toronto   
13     13         M5A  Downtown Toronto   
14     14         M5B  Downtown Toronto   
15     15         M5C  Downtown Toronto   
16     16         M5E  Downtown Toronto   
17     17         M5G  Downtown Toronto   
18     18         M5H  Downtown Toronto   
19     19         M5J  Downtown Toronto   
20     20         M5K  Downtown Toronto   
21     21         M5L  Downtown Toronto   
22     22         M5N   Central Toronto   
23     23         M5P   Central Toronto   
24     24         M5R   Central Toronto   
25     25         M5S  Downtown Toronto   
26     26         M5T  Downtown Toronto   
27     27         M5V  Downtown Toronto   
28     28         M5W  Downtown Toronto   
29     29         M5X  Downtown Toronto   
30     30         M6G  Downtown Toronto   
31     31         M6H      West Toronto   
32     32         M6J      West Toronto   
33     33         M6K      West Toronto   
34     34         M6P      West Toronto   
35     35         M6R      West Toronto   
36     36         M6S      West Toronto   
37     37         M7A  Downtown Toronto   
38     38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      India Bazaar, The Beaches West  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                   North Toronto West, Lawrence Park  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        St. James Town, Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Regent Park, Harbourfront  43.654260 -79.360636  
14                           Garden District, Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Richmond, Adelaide, King  43.650571 -79.384568  
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     University of Toronto, Harbord  43.662696 -79.400049  
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
28                   

In [124]:
toronto_data = toronto_data.drop('index', 1)

toronto_data

Postal Code           Borough  \
0          M4E      East Toronto   
1          M4K      East Toronto   
2          M4L      East Toronto   
3          M4M      East Toronto   
4          M4N   Central Toronto   
5          M4P   Central Toronto   
6          M4R   Central Toronto   
7          M4S   Central Toronto   
8          M4T   Central Toronto   
9          M4V   Central Toronto   
10         M4W  Downtown Toronto   
11         M4X  Downtown Toronto   
12         M4Y  Downtown Toronto   
13         M5A  Downtown Toronto   
14         M5B  Downtown Toronto   
15         M5C  Downtown Toronto   
16         M5E  Downtown Toronto   
17         M5G  Downtown Toronto   
18         M5H  Downtown Toronto   
19         M5J  Downtown Toronto   
20         M5K  Downtown Toronto   
21         M5L  Downtown Toronto   
22         M5N   Central Toronto   
23         M5P   Central Toronto   
24         M5R   Central Toronto   
25         M5S  Downtown Toronto   
26         M5T  Downtown Toronto   
27         M5V  Downtown Toronto   
28         M5W  Downtown Toronto   
29         M5X  Downtown Toronto   
30         M6G  Downtown Toronto   
31         M6H      West Toronto   
32         M6J      West Toronto   
33         M6K      West Toronto   
34         M6P      West Toronto   
35         M6R      West Toronto   
36         M6S      West Toronto   
37         M7A  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      India Bazaar, The Beaches West  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                   North Toronto West, Lawrence Park  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        St. James Town, Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Regent Park, Harbourfront  43.654260 -79.360636  
14                           Garden District, Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Richmond, Adelaide, King  43.650571 -79.384568  
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     University of Toronto, Harbord  43.662696 -79.400049  
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
28                                     Stn A PO Boxes  43.646435 -79.374846  
29             First Canadian Place, Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                       Dufferin, Dovercourt Village  43.669005 -

Get the neighborhood's name.

In [125]:
toronto_data.loc[0, 'Neighborhood']

'The Beaches'

Get the neighborhood's latitude and longitude values.

In [126]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [128]:
LIMIT = 100

radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


'https://api.foursquare.com/v2/venues/explore?&client_id=MLTOHZTUFRKGHKUQR4O2PMTX3ANCOZ3XMLSUC2VNFJMISJ4X&client_secret=HOHDJNMA5NRIFF523PDNQX2LDE2G24VHYC4HBRMFAO0H51SQ&v=20180605&ll=43.67635739999999,-79.2930312&radius=500&limit=100'

Send the GET request and examine the results

In [129]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ede9a7dc8cff2001bf1e1ea'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'labe

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [130]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [132]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(10)

name         categories        lat        lng
0                   Glen Manor Ravine              Trail  43.676821 -79.293942
1  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
2                 Grover Pub and Grub                Pub  43.679181 -79.297215
3                       Upper Beaches       Neighborhood  43.680563 -79.292869

I realize at this point that there are only 4 venues that came back for this neighborhood. I'm curious if Toronto in general is more spread out with less venues, so perhaps we should expand with a radius larger than 500 meters. I will use 1,000:

In [145]:
LIMIT = 100

radius = 1000

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


'https://api.foursquare.com/v2/venues/explore?&client_id=MLTOHZTUFRKGHKUQR4O2PMTX3ANCOZ3XMLSUC2VNFJMISJ4X&client_secret=HOHDJNMA5NRIFF523PDNQX2LDE2G24VHYC4HBRMFAO0H51SQ&v=20180605&ll=43.67635739999999,-79.2930312&radius=1000&limit=100'

In [146]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5edea097006dce001b89c5c1'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 75,
  'suggestedBounds': {'ne': {'lat': 43.685357409000005,
    'lng': -79.28061062898105},
   'sw': {'lat': 43.66735739099998, 'lng': -79.30545177101895}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'lab

In [147]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [148]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(10)

name                     categories        lat        lng
0  Glen Manor Ravine                          Trail  43.676821 -79.293942
1    Tori's Bakeshop  Vegetarian / Vegan Restaurant  43.672114 -79.290331
2  Beaches Bake Shop                         Bakery  43.680363 -79.289692
3     The Beech Tree                      Gastropub  43.680493 -79.288846
4    The Fox Theatre            Indie Movie Theater  43.672801 -79.287272
5    Ed's Real Scoop                 Ice Cream Shop  43.672630 -79.287993
6     Bagels On Fire                     Bagel Shop  43.672864 -79.286784
7    Mastermind Toys               Toy / Game Store  43.671453 -79.293971
8     Veloute Bistro              French Restaurant  43.672267 -79.289584
9       Beacher Cafe                 Breakfast Spot  43.671938 -79.291238

In [149]:
nearby_venues.shape

(75, 4)

Interesting. Going from radius of 500 to 1,000 increases our response from 4 to 75. I'll leave it at 1,000 radius for now

In [150]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

75 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Toronto

Let's create a function to repeat the same process to all the neighborhoods in Toronto, using 1,000 meters as the radius as we did above

In [151]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [154]:
Toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

#### Let's check the size of the resulting dataframe

In [156]:
print(Toronto_venues.shape)
Toronto_venues.head()

(3180, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

               Venue  Venue Latitude  Venue Longitude  \
0  Glen Manor Ravine       43.676821       -79.293942   
1    Tori's Bakeshop       43.672114       -79.290331   
2  Beaches Bake Shop       43.680363       -79.289692   
3     The Beech Tree       43.680493       -79.288846   
4    The Fox Theatre       43.672801       -79.287272   

                  Venue Category  
0                          Trail  
1  Vegetarian / Vegan Restaurant  
2                         Bakery  
3                      Gastropub  
4            Indie Movie Theater

Let's check how many venues were returned for each neighborhood

In [157]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                           100   
Brockton, Parkdale Village, Exhibition Place                          100   
Business reply mail Processing Centre, South Ce...                     47   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                    100   
Christie                                                              100   
Church and Wellesley                                                  100   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                            100   
Davisville North                                                      100   
Dufferin, Dovercourt Village                                           67   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                        48   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                         100   
India Bazaar, The Beaches West                                         81   
Kensington Market, Chinatown, Grange Park                             100   
Lawrence Park                                                           9   
Little Portugal, Trinity                                              100   
Moore Park, Summerhill East                                            64   
North Toronto West, Lawrence Park                                      45   
Parkdale, Roncesvalles                                                100   
Queen's Park, Ontario Provincial Government                           100   
Regent Park, Harbourfront                                             100   
Richmond, Adelaide, King                                              100   
Rosedale                                                               21   
Roselawn                                                               21   
Runnymede, Swansea                                                     78   
St. James Town                                                        100   
St. James Town, Cabbagetown                                            34   
Stn A PO Boxes                                                        100   
Studio District                                                       100   
Summerhill West, Rathnelly, South Hill, Forest ...                     75   
The Annex, North Midtown, Yorkville                                   100   
The Beaches                                                            75   
The Danforth West, Riverdale                                          100   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                        100   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                            100   
Brockton, Parkdale Village, Exhibition Place                           100   
Business reply mail Processing Centre, South Ce...                      47   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                     100   
Christie                                                               100   
Church and Wellesley                                                   100   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                             100   
Davisville North                    

At this point I'm glad I increased the radius. Our 3,180 venues is really close to the 3,122 in the New York exercise and without changing it, many neighborhoods wouldn't ultimately be able to display a top 10

#### Let's find out how many unique categories can be curated from all the returned venues

In [158]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 273 uniques categories.


## 3. Analyze Each Neighborhood

In [159]:
# one hot encoding
toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Zoo  Accessories Store  Airport  Airport Lounge  American Restaurant  \
0    0                  0        0               0                    0   
1    0                  0        0               0                    0   
2    0                  0        0               0                    0   
3    0                  0        0               0                    0   
4    0                  0        0               0                    0   

   Amphitheater  Animal Shelter  Antique Shop  Aquarium  Art Gallery  \
0             0               0             0         0            0   
1             0               0             0         0            0   
2             0               0             0         0            0   
3             0               0             0         0            0   
4             0               0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Dealership  Auto Workshop  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  \
0                0              0          0           0       0     0    0   
1                0              0          0           0       0     0    0   
2                0              0          0           0       1     0    0   
3                0              0          0           0       0     0    0   
4                0              0          0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  Beach Bar  \
0               0                 0                   0      0          0   
1               0                 0                   0      0          0   
2               0                 0                   0      0          0   
3               0                 0                   0      0          0   
4               0                 0                   0      0          0   

   Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  Bookstore  \
0         0           0                   0          0       0          0   
1         0           0                   0          0       0          0   
2         0           0                   0          0       0          0   
3         0           0                   0          0       0          0   
4         0           0                   0          0       0          0   

   Botanical Garden  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0                 0         0                     0               0        0   
1                 0         0                     0               0        0   
2                 0         0                     0               0        0   
3                 0         0                     0               0        0   
4                 0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  \
0            0                0       0             0              0   
1            0                0       0             0              0   
2            0                0       0             0              0   
3            0                0       0             0              0   
4            0                0       0             0              0   

   Bus Stop  Business Service  Butcher  Café  Cajun / Creole Restaurant  \
0         0                 0        0     0                          0   
1         0                 0        0     0                          0   
2         0                 0        0     0                          0   
3         0                 0        0     0   

And let's examine the new dataframe size.

In [160]:
toronto_onehot.shape

(3180, 273)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [161]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood   Zoo  \
0                                         Berczy Park  0.00   
1        Brockton, Parkdale Village, Exhibition Place  0.00   
2   Business reply mail Processing Centre, South C...  0.00   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.00   
4                                  Central Bay Street  0.00   
5                                            Christie  0.00   
6                                Church and Wellesley  0.00   
7                      Commerce Court, Victoria Hotel  0.00   
8                                          Davisville  0.00   
9                                    Davisville North  0.00   
10                       Dufferin, Dovercourt Village  0.00   
11             First Canadian Place, Underground city  0.00   
12    Forest Hill North & West, Forest Hill Road Park  0.00   
13                           Garden District, Ryerson  0.00   
14  Harbourfront East, Union Station, Toronto Islands  0.00   
15                      High Park, The Junction South  0.00   
16                     India Bazaar, The Beaches West  0.00   
17          Kensington Market, Chinatown, Grange Park  0.00   
18                                      Lawrence Park  0.00   
19                           Little Portugal, Trinity  0.00   
20                        Moore Park, Summerhill East  0.00   
21                  North Toronto West, Lawrence Park  0.00   
22                             Parkdale, Roncesvalles  0.01   
23        Queen's Park, Ontario Provincial Government  0.00   
24                          Regent Park, Harbourfront  0.00   
25                           Richmond, Adelaide, King  0.00   
26                                           Rosedale  0.00   
27                                           Roselawn  0.00   
28                                 Runnymede, Swansea  0.00   
29                                     St. James Town  0.00   
30                        St. James Town, Cabbagetown  0.00   
31                                     Stn A PO Boxes  0.00   
32                                    Studio District  0.00   
33  Summerhill West, Rathnelly, South Hill, Forest...  0.00   
34                The Annex, North Midtown, Yorkville  0.00   
35                                        The Beaches  0.00   
36                       The Danforth West, Riverdale  0.00   
37           Toronto Dominion Centre, Design Exchange  0.00   
38                     University of Toronto, Harbord  0.00   

    Accessories Store   Airport  Airport Lounge  American Restaurant  \
0                0.00  0.000000        0.000000             0.010000   
1                0.01  0.000000        0.000000             0.010000   
2                0.00  0.000000        0.000000             0.021277   
3                0.00  0.066667        0.066667             0.000000   
4                0.00  0.000000        0.000000             0.010000   
5                0.00  0.000000        0.000000             0.010000   
6                0.00  0.000000        0.000000             0.010000   
7                0.00  0.000000        0.000000             0.020000   
8                0.00  0.000000        0.000000             0.000000   
9                0.00  0.000000        0.000000             0.000000   
10               0.00  0.000000        0.000000             0.000000   
11               0.00  0.000000        0.000000             0.020000   
12               0.00  0.000000        0.000000             0.000000   
13               0.00  0.000000        0.000000             0.010000   
14               0.00  0.000000        0.000000             0.000000   
15               0.00  0.000000        0.000000             0.000000   
16               0.00  0.000000        0.000000             0.000000   
17               0.00  0.000000        0.000000             0.000000   
18               0.00  0.000000        0.000000             0.000000   
19               0.00  0.000000        0.000000             0.010000   
20      

#### Let's confirm the new size

In [162]:
toronto_grouped.shape

(39, 273)

#### Let's print each neighborhood along with the top 5 most common venues

In [163]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                 venue  freq
0          Coffee Shop  0.12
1                 Café  0.07
2                Hotel  0.05
3  Japanese Restaurant  0.04
4           Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0                    Café  0.07
1             Coffee Shop  0.06
2              Restaurant  0.05
3                     Bar  0.05
4  Furniture / Home Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
              venue  freq
0              Park  0.09
1       Coffee Shop  0.06
2       Pizza Place  0.06
3           Brewery  0.06
4  Sushi Restaurant  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0      Coffee Shop  0.13
1             Café  0.13
2  Harbor / Marina  0.13
3             Park  0.07
4          Airport  0.07


----Central Bay Street----
            

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [164]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [166]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...                  Café   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                 Hotel            Restaurant   
1           Coffee Shop            Restaurant                   Bar   
2               Brewery           Pizza Place           Coffee Shop   
3       Harbor / Marina           Coffee Shop        Scenic Lookout   
4                  Café      Sushi Restaurant                  Park   

  5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0   Japanese Restaurant                    Park          Liquor Store   
1                Bakery  Furniture / Home Store             Gift Shop   
2  Fast Food Restaurant        Sushi Restaurant    Italian Restaurant   
3                Garden                 Airport        Airport Lounge   
4        Clothing Store        Ramen Restaurant           Art Gallery   

    8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0               Gastropub          Cocktail Bar                  Bakery  
1      Tibetan Restaurant           Supermarket  Thrift / Vintage Store  
2                     Pub          Liquor Store              Comic Shop  
3                 Dog Run      Sculpture Garden                    Park  
4  Furniture / Home Store                 Hotel                 Theater

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [168]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 0, 3, 4, 0, 0, 0, 2, 0, 3], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [172]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels New', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code          Borough                    Neighborhood   Latitude  \
0         M4E     East Toronto                     The Beaches  43.676357   
1         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3         M4M     East Toronto                 Studio District  43.659526   
4         M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels New  Cluster Labels 1st Most Common Venue  \
0 -79.293031                   0               0                   Pub   
1 -79.352188                   0               0      Greek Restaurant   
2 -79.315572                   0               0     Indian Restaurant   
3 -79.340923                   0               0           Coffee Shop   
4 -79.388790                   1               1            Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop           Pizza Place                 Beach   
1           Coffee Shop                  Café                   Pub   
2           Coffee Shop  Fast Food Restaurant                  Park   
3                   Bar                  Café               Brewery   
4             Bookstore  Gym / Fitness Center          College Quad   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Park   Japanese Restaurant        Breakfast Spot   
1    Italian Restaurant  Fast Food Restaurant           Pizza Place   
2                  Café                 Beach         Burrito Place   
3   American Restaurant                Bakery                 Diner   
4           College Gym           Coffee Shop                  Café   

   8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0         Sandwich Place          Burger Joint                     Bar  
1                   Bank                Bakery  Furniture / Home Store  
2                 Bakery    Italian Restaurant          Sandwich Place  
3  Vietnamese Restaurant     French Restaurant      Italian Restaurant  
4                  Trail                  Park     Dumpling Restaurant

Finally, let's visualize the resulting clusters

In [173]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

In [178]:
toronto_merged.loc[toronto_merged['Cluster Labels New'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels New  Cluster Labels  \
0       East Toronto                   0               0   
1       East Toronto                   0               0   
2       East Toronto                   0               0   
3       East Toronto                   0               0   
6    Central Toronto                   0               0   
7    Central Toronto                   0               0   
11  Downtown Toronto                   0               0   
12  Downtown Toronto                   0               0   
17  Downtown Toronto                   0               0   
24   Central Toronto                   0               0   
25  Downtown Toronto                   0               0   
26  Downtown Toronto                   0               0   
30  Downtown Toronto                   0               0   
32      West Toronto                   0               0   
33      West Toronto                   0               0   
34      West Toronto                   0               0   
35      West Toronto                   0               0   
37  Downtown Toronto                   0               0   

   1st Most Common Venue          2nd Most Common Venue  \
0                    Pub                    Coffee Shop   
1       Greek Restaurant                    Coffee Shop   
2      Indian Restaurant                    Coffee Shop   
3            Coffee Shop                            Bar   
6            Coffee Shop             Italian Restaurant   
7     Italian Restaurant                    Coffee Shop   
11                 Diner            Japanese Restaurant   
12           Coffee Shop            Japanese Restaurant   
17           Coffee Shop                           Café   
24                  Café  Vegetarian / Vegan Restaurant   
25                  Café                         Bakery   
26                  Café                            Bar   
30     Korean Restaurant                    Coffee Shop   
32                  Café                            Bar   
33                  Café                    Coffee Shop   
34                  Café                            Bar   
35           Pizza Place                    Coffee Shop   
37           Coffee Shop               Sushi Restaurant   

            3rd Most Common Venue          4th Most Common Venue  \
0                     Pizza Place                          Beach   
1                            Café                            Pub   
2            Fast Food Restaurant                           Park   
3                            Café                        Brewery   
6              Mexican Restaurant                          Diner   
7                Sushi Restaurant                    Pizza Place   
11                           Park                     Restaurant   
12                     Restaurant               Sushi Restaurant   
17               Sushi Restaurant                           Park   
24                    Coffee Shop             Italian Restaurant   
25                            Bar                    Coffee Shop   
26  Vegetarian / Vegan Restaurant                    Art Gallery   
30                           Café                  Grocery Store   
32                     Restaurant  Vegetarian / Vegan Restaurant   
33                     Restaurant                            Bar   
34                    Coffee Shop              Convenience Store   
35                            Bar                         Bakery   
37                           Park               Ramen Restaurant   

   5th Most Common Venue   6th Most Common Venue  \
0                   Park     Japanese Restaurant   
1     Italian Restaurant    Fast Food Restaurant   
2                   Café                   Beach   
3    American Restaurant                  Bakery   
6           Skating Rink                    Café   
7                   Café                     Gym   
11                  Café               Gastropub   
12                  Park            Burge

In [179]:
toronto_merged.loc[toronto_merged['Cluster Labels New'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels New  Cluster Labels 1st Most Common Venue  \
4  Central Toronto                   1               1            Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4             Bookstore  Gym / Fitness Center          College Quad   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4           College Gym           Coffee Shop                  Café   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4                 Trail                  Park    Dumpling Restaurant

In [180]:
toronto_merged.loc[toronto_merged['Cluster Labels New'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels New  Cluster Labels  \
13  Downtown Toronto                   2               2   
14  Downtown Toronto                   2               2   
15  Downtown Toronto                   2               2   
16  Downtown Toronto                   2               2   
18  Downtown Toronto                   2               2   
19  Downtown Toronto                   2               2   
20  Downtown Toronto                   2               2   
21  Downtown Toronto                   2               2   
28  Downtown Toronto                   2               2   
29  Downtown Toronto                   2               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
13           Coffee Shop                   Pub                  Park   
14           Coffee Shop             Gastropub   Japanese Restaurant   
15           Coffee Shop                  Café            Restaurant   
16           Coffee Shop                  Café                 Hotel   
18           Coffee Shop                  Café                 Hotel   
19           Coffee Shop                  Café                 Hotel   
20           Coffee Shop                 Hotel                  Café   
21           Coffee Shop                  Café                 Hotel   
28           Coffee Shop                  Café            Restaurant   
29           Coffee Shop                 Hotel                  Café   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
13                  Café                 Diner               Theater   
14               Theater            Restaurant                 Hotel   
15             Gastropub    Italian Restaurant               Theater   
16            Restaurant   Japanese Restaurant                  Park   
18               Theater   Japanese Restaurant              Tea Room   
19                  Park   Japanese Restaurant               Brewery   
20   Japanese Restaurant            Restaurant          Concert Hall   
21            Restaurant   Japanese Restaurant               Theater   
28                 Hotel   Japanese Restaurant                  Park   
29               Theater          Concert Hall            Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
13        Breakfast Spot            Restaurant                Bakery   
14    Italian Restaurant                 Diner         Burrito Place   
15          Cocktail Bar    Seafood Restaurant                 Plaza   
16          Liquor Store             Gastropub          Cocktail Bar   
18      Sushi Restaurant            Restaurant                   Gym   
19               Theater                   Gym                Bakery   
20               Theater          Cocktail Bar                   Gym   
21             Gastropub    Seafood Restaurant                Bakery   
28             Gastropub    Seafood Restaurant          Cocktail Bar   
29   Japanese Restaurant                 Plaza         Deli / Bodega   

   10th Most Common Venue  
13     Italian Restaurant  
14       Ramen Restaurant  
15                  Hotel  
16                 Bakery  
18            Pizza Place  
19                  Plaza  
20                  Plaza  
21           Concert Hall  
28            Art Gallery  
29     Italian Restaurant

In [181]:
toronto_merged.loc[toronto_merged['Cluster Labels New'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels New  Cluster Labels  \
5    Central Toronto                   3               3   
8    Central Toronto                   3               3   
9    Central Toronto                   3               3   
10  Downtown Toronto                   3               3   
22   Central Toronto                   3               3   
23   Central Toronto                   3               3   
31      West Toronto                   3               3   
36      West Toronto                   3               3   
38      East Toronto                   3               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
5            Coffee Shop    Italian Restaurant            Restaurant   
8          Grocery Store    Italian Restaurant           Coffee Shop   
9            Coffee Shop      Sushi Restaurant                  Park   
10           Coffee Shop                  Park         Grocery Store   
22                  Bank      Sushi Restaurant              Pharmacy   
23           Coffee Shop                  Café                  Bank   
31                  Café                  Park           Coffee Shop   
36           Coffee Shop                  Café                Bakery   
38                  Park               Brewery           Pizza Place   

   4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
5                   Café                    Gym  Fast Food Restaurant   
8            Pizza Place        Thai Restaurant                  Park   
9     Italian Restaurant        Thai Restaurant         Grocery Store   
10        Breakfast Spot     Athletics & Sports            Playground   
22    Italian Restaurant            Coffee Shop             Gastropub   
23                  Park   Gym / Fitness Center    Italian Restaurant   
31      Sushi Restaurant  Portuguese Restaurant                   Bar   
36           Pizza Place       Sushi Restaurant                   Pub   
38           Coffee Shop   Fast Food Restaurant      Sushi Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5           Dessert Shop      Sushi Restaurant           Pizza Place   
8             Restaurant                  Café                   Gym   
9             Restaurant          Liquor Store           Pizza Place   
10        Sandwich Place           Candy Store         Metro Station   
22           Dry Cleaner                Bakery            Bagel Shop   
23   Japanese Restaurant      Sushi Restaurant          Burger Joint   
31                Bakery          Gourmet Shop              Pharmacy   
36    Italian Restaurant    Falafel Restaurant           Flower Shop   
38    Italian Restaurant                   Pub          Liquor Store   

   10th Most Common Venue  
5             Yoga Studio  
8                     Pub  
9              Bagel Shop  
10    Filipino Restaurant  
22    Japanese Restaurant  
23           Skating Rink  
31     Italian Restaurant  
36                   Bank  
38             Comic Shop

In [182]:
toronto_merged.loc[toronto_merged['Cluster Labels New'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels New  Cluster Labels  \
27  Downtown Toronto                   4               4   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
27                  Café       Harbor / Marina           Coffee Shop   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
27        Scenic Lookout                Garden               Airport   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
27        Airport Lounge               Dog Run      Sculpture Garden   

   10th Most Common Venue  
27                   Park

# And that's it!!!!!!